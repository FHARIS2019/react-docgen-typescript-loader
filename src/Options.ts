import * as Ajv from "ajv";

export interface Options {
  includes: RegExp[];
  excludes: RegExp[];
  docgenCollectionName: string;
  verbose: boolean;
}

const ajv = new Ajv();

const validate = ajv.compile({
  type: "object",
  properties: {
    includes: {
      type: "array",
      items: {
        type: "object",
      },
    },
    excludes: {
      type: "array",
      items: {
        type: "object",
      },
    },
    docgenCollectionName: {
      type: "string",
    },
    verbose: {
      type: "boolean",
    },
  },
  additionalProperties: false,
});

export function validateOptions(options?: object) {
  if (options === undefined) return;

  const valid = validate(options);

  if (!valid) {
    const errorObj = validate.errors![0];
    let message = "react-docgen-typescript-webpack-plugin: ";
    if (errorObj.dataPath) message += errorObj.dataPath + " ";
    message += errorObj.message;

    throw new Error(message);
  }
}
