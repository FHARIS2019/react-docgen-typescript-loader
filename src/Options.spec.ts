import { validateOptions } from "./Options";

describe("validateOptions", () => {
  it("does not error on undefined", () => {
    validateOptions();
  });

  it("throws on non-object", () => {
    expect(() => validateOptions("fail" as any)).toThrowErrorMatchingSnapshot();
  });

  it("throws on unexpected property", () => {
    expect(() =>
      validateOptions({ unexpected: "" }),
    ).toThrowErrorMatchingSnapshot();
  });

  it("throws on include not being array of strings", () => {
    expect(() =>
      validateOptions({ includes: "fail" }),
    ).toThrowErrorMatchingSnapshot();

    expect(() => {
      validateOptions({ includes: [3] });
    }).toThrowErrorMatchingSnapshot();
  });

  it("does not throw on empty object", () => {
    validateOptions({});
  });
});
