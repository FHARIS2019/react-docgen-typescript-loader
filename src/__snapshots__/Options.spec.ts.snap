// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateOptions throws on include not being array of strings 1`] = `"react-docgen-typescript-webpack-plugin: .includes should be array"`;

exports[`validateOptions throws on include not being array of strings 2`] = `"react-docgen-typescript-webpack-plugin: .includes[0] should be object"`;

exports[`validateOptions throws on non-object 1`] = `"react-docgen-typescript-webpack-plugin: should be object"`;

exports[`validateOptions throws on unexpected property 1`] = `"react-docgen-typescript-webpack-plugin: should NOT have additional properties"`;
